### PR TITLE
fix(encoding/toml): fix inline table and nested array

### LINKED ISSUE #823: TOML parser's inline/nested tables misaligned with real TOML
I'm comparing this TOML parse with http://toml-online-parser.ovonick.com

First opened here because I'm dumb: https://github.com/denoland/deno/issues/9919

## case 1

This is proper TOML (AFAIK) that parses correctly elsewhere:

```toml
annotation_filter = { "kubernetes.io/ingress.class" = "nginx" }
```

... but Deno's parser tries reading as JSON:

```
> TOML.parse(`annotation_filter = { "kubernetes.io/ingress.class" = "nginx" }`);
error: Uncaught SyntaxError: Unexpected string in JSON at position 31
      return JSON.parse(dataString);
                  ^
    at JSON.parse (<anonymous>)
    at Parser._parseInlineTableOrArray (https://deno.land/std@0.91.0/encoding/toml.ts:322:19)
    at Parser._parseData (https://deno.land/std@0.91.0/encoding/toml.ts:275:21)
    at Parser._processDeclaration (https://deno.land/std@0.91.0/encoding/toml.ts:264:24)
    at Parser._parseLines (https://deno.land/std@0.91.0/encoding/toml.ts:492:25)
    at Parser.parse (https://deno.land/std@0.91.0/encoding/toml.ts:541:10)
    at Module.parse (https://deno.land/std@0.91.0/encoding/toml.ts:735:33)
```

## case 2

As a work around, I tried this non-inline syntax instead:

```toml
[[aaa]]
hi = "hi"
[aaa.bbb]
asdf = "asdf"
```

... but the key just silently gets lost:

```
> TOML.parse(`[[aaa]]\nhi = "hi"\n[aaa.bbb]\nasdf = "asdf"`)
{ aaa: [ { hi: "hi" } ] }
```

I was expecting:

```json
{ "aaa": [ {
            "hi": "hi",
            "bbb": {
                "asdf": "asdf"
            }
} ] }
```

## case 3

This actually works with Deno as of std@0.91.0:
```toml
[[source]]
annotation_filter = { "\"kubernetes.io/ingress.class\"": "nginx" }
 # also this comment is read as a key because of leading spaces = "..."
```

That last line seems to be included because of some parsing bug:

```
> TOML.parse(`[[source]]\nannotation_filter = { "\\"kubernetes.io/ingress.class\\"": "nginx" }\n # also this comment is read as a key because of leading spaces = "..."`)
{
  source: [
    {
      annotation_filter: { "kubernetes.io/ingress.class": "nginx" },
      "# also this comment is read as a key because of leading spaces": "..."
    }
  ]
}
```

**That inline table is not valid TOML** and yet it parses ok so I've been using it anyway 😒

### PR DESCRIPTION
**Pull request by FujiHaruka**
close #823

Sorry for a large change. I almost overall rewrite TOML parser based on recursive descent parsing because current parser includes many edge case bugs. This will make the parser more stable and maintainable.

Of course all tests passed, by which we can believe the parser doesn't get worse. I also added some test cases the previous parser doesn't pass.

This PR fixes case 1, 2, 3, and 4 of https://github.com/denoland/deno_std/issues/823#issuecomment-814367934.

In addition, if the parser fails to parse, it outputs the position the error occurred like this:

> Parse error on line 1, column 34: Single-line string cannot contain EOL

### CODE REVIEW

File: encoding/_toml/parser.ts [hunk 1]
```diff
@@ -0,0 +1,897 @@
+// Copyright 2018-2021 the Deno authors. All rights reserved. MIT license.
+import { deepAssign } from "https://deno.land/std@0.101.0/_util/deep_assign.ts";
```
Comment (kt3k) [inline encoding/_toml/parser.ts:2]: Please use relative paths instead of versioned std module url (to prevent multiple versions of std mixed in the future)

File: encoding/_toml/parser_test.ts [hunk 2]
```diff
@@ -0,0 +1,441 @@
+// Copyright 2018-2021 the Deno authors. All rights reserved. MIT license.
+import {
+  assertEquals,
+  assertThrows,
+} from "https://deno.land/std@0.101.0/testing/asserts.ts";
```
Comment (kt3k) [inline encoding/_toml/parser_test.ts:5]: This also needs to be relative path dependency.

**Comment by kt3k** [comment]
I think this is well designed rewrite of TOML parser. The unit tests of each parser component look great to me. Those make the maintenance of this module significantly easier in my view. Thank you for the contribution.

File: encoding/_toml/parser.ts (hunk 1 above)
Comment (FujiHaruka) [reply encoding/_toml/parser.ts]: Oh, I didn't notice that. Thank you. I fixed!

File: encoding/_toml/parser_test.ts (hunk 2 above)
Comment (FujiHaruka) [reply encoding/_toml/parser_test.ts]: And fixed too!

**Review by ry — APPROVED** [review]
LGTM - huge patch but seems quite clean and well implemented. Thank you!
